src/ReplaceReferencesCapableTrait.php

```diff
@@ -0,0 +1,76 @@
+<?php
+
+namespace Dhii\Config;
+
+use InvalidArgumentException;
+use Psr\Container\ContainerExceptionInterface;
+use Psr\Container\ContainerInterface;
+use Dhii\Util\String\StringableInterface as Stringable;
+use Psr\Container\NotFoundExceptionInterface;
+
+/**
+ * Trait ReplaceReferencesCapableTrait.
+ *
+ * Adds ability to replace all tokens in the form `${key}` in a string with corresponding values retrieved from a container.
+ *
+ * @since [*next-version*]
+ */
+trait ReplaceReferencesCapableTrait
+{
+    /**
+     * Replaces all tokens in the form `${key}` in a string with corresponding values retrieved from a container.
+     *
+     * @since [*next-version*]
+     *
+     * @param string|Stringable  $input          Input string to find and replace references
+     * @param ContainerInterface $container
+     * @param string|Stringable  $startDelimiter
+     * @param mixed              $default        String to replace reference if key not found in container.
+     * @param string|Stringable  $endDelimiter
+     *
+     * @throws ContainerExceptionInterface
+     * @throws NotFoundExceptionInterface
+     *
+     * @return string The resulting string.
+     */
+    protected function _replaceReferences($input, ContainerInterface $container, $default = null, $startDelimiter = '${', $endDelimiter = '}')
+    {
+        $input        = $this->_normalizeString($input);
+        $defaultValue = $default === null ? '' : $this->_normalizeString($default);
+
+        $startDelimiter = preg_quote($this->_normalizeString($startDelimiter));
```

Comment:
In my earlier comment, I mentioned that you know the RegEx delimiter. For some reason, you didn't use it here, which means that it will not be escaped.